File: src/simple_agent/llm.py
from __future__ import annotations

from dataclasses import dataclass
import itertools
import re
from typing import Any, Protocol


Message = dict[str, Any]


@dataclass(slots=True)
class ToolSpec:
    name: str
    description: str
    input_schema: dict[str, Any]


class ModelProvider(Protocol):
    def generate(self, messages: list[Message], tools: list[ToolSpec]) -> Message:
        """
        Returns an assistant message shaped like pi-mono AssistantMessage:
        {
          "role": "assistant",
          "content": [{"type":"text","text":"..."}, {"type":"toolCall",...}],
          "stopReason": "complete" | "error" | "aborted"
        }
        """


class MockProvider:
    """Deterministic local provider that emits pi-mono-like assistant messages."""

    def __init__(self) -> None:
        self._counter = itertools.count(1)

    def generate(self, messages: list[Message], tools: list[ToolSpec]) -> Message:
        latest = messages[-1] if messages else {"role": "user", "content": ""}
        latest_role = str(latest.get("role", "user"))
        latest_text = _extract_text(latest).strip().lower()

        if latest_role == "toolResult":
            return _assistant_text(f"Tool result:\n{_extract_text(latest)}")

        read_match = re.search(r"read(?: file)?\s+(.+)$", latest_text)
        if read_match and _has_tool(tools, "read"):
            return _assistant_with_tool_call(
                text="I'll read that file.",
                name="read",
                arguments={"path": read_match.group(1).strip().strip("'\"")},
                call_id=f"call_{next(self._counter)}",
            )

        ls_match = re.search(r"(?:^|\s)(ls|list files|show files)(?:$|\s)", latest_text)
        if ls_match and _has_tool(tools, "ls"):
            return _assistant_with_tool_call(
                text="I'll list the directory.",
                name="ls",
                arguments={"path": "."},
                call_id=f"call_{next(self._counter)}",
            )

        bash_match = re.search(r"(?:run command|bash)\s+(.+)$", latest_text)
        if bash_match and _has_tool(tools, "bash"):
            return _assistant_with_tool_call(
                text="I'll run that command.",
                name="bash",
                arguments={"command": bash_match.group(1).strip().strip("'\"")},
                call_id=f"call_{next(self._counter)}",
            )

        names = ", ".join(spec.name for spec in tools)
        return _assistant_text(
            "Mock provider active. Try: 'ls', 'read <path>', or 'bash <command>'. "
            f"Available tools: {names}."
        )


def _assistant_text(text: str) -> Message:
    return {
        "role": "assistant",
        "content": [{"type": "text", "text": text}],
        "stopReason": "complete",
        "timestamp": _now_ms(),
    }


def _assistant_with_tool_call(text: str, name: str, arguments: dict[str, Any], call_id: str) -> Message:
    return {
        "role": "assistant",
        "content": [
            {"type": "text", "text": text},
            {"type": "toolCall", "id": call_id, "name": name, "arguments": arguments},
        ],
        "stopReason": "complete",
        "timestamp": _now_ms(),
    }


def _extract_text(message: Message) -> str:
    content = message.get("content")
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        texts: list[str] = []
        for item in content:
            if isinstance(item, dict) and item.get("type") == "text":
                texts.append(str(item.get("text", "")))
        return "\n".join(t for t in texts if t)
    return str(content or "")


def _has_tool(tools: list[ToolSpec], name: str) -> bool:
    return any(tool.name == name for tool in tools)


def _now_ms() -> int:
    import time

    return int(time.time() * 1000)

File: src/llm.rs
#[derive(Clone, Debug)]
pub struct ToolArgSpec {
    pub name: String,
    pub description: String,
    pub required: bool,
}

#[derive(Clone, Debug)]
pub struct ToolSpec {
    pub name: String,
    pub description: String,
    pub args: Vec<ToolArgSpec>,
}

#[derive(Clone, Debug)]
pub enum ContentItem {
    Text(String),
    ToolCall {
        id: String,
        name: String,
        arguments: Vec<(String, String)>,
    },
}

#[derive(Clone, Debug)]
pub struct Message {
    pub role: String,
    pub content: Vec<ContentItem>,
}

pub trait ModelProvider {
    fn generate(&mut self, messages: &[Message], tools: &[ToolSpec]) -> Message;
}

pub struct MockProvider {
    counter: usize,
}

impl MockProvider {
    pub fn new() -> Self {
        Self { counter: 0 }
    }
}

impl ModelProvider for MockProvider {
    fn generate(&mut self, messages: &[Message], tools: &[ToolSpec]) -> Message {
        let latest = messages.last();
        let role = latest.map(|m| m.role.as_str()).unwrap_or("user");
        let text = latest.map(extract_text).unwrap_or_default().to_lowercase();

        if role == "toolResult" {
            return Message {
                role: "assistant".to_string(),
                content: vec![ContentItem::Text(format!(
                    "Tool result:\n{}",
                    latest.map(extract_text).unwrap_or_default()
                ))],
            };
        }

        if text.starts_with("read ") && has_tool(tools, "read") {
            let path = text.replacen("read ", "", 1);
            return self.tool_call(
                "I'll read that file.",
                "read",
                vec![("path".to_string(), path)],
            );
        }
        if text.contains("ls") && has_tool(tools, "ls") {
            return self.tool_call(
                "I'll list the directory.",
                "ls",
                vec![("path".to_string(), ".".to_string())],
            );
        }
        if text.starts_with("bash ") && has_tool(tools, "bash") {
            let cmd = text.replacen("bash ", "", 1);
            return self.tool_call(
                "I'll run that command.",
                "bash",
                vec![("command".to_string(), cmd)],
            );
        }

        let names = tools
            .iter()
            .map(|t| t.name.clone())
            .collect::<Vec<_>>()
            .join(", ");
        Message {
            role: "assistant".to_string(),
            content: vec![ContentItem::Text(format!(
                "Mock provider active. Try: 'ls', 'read <path>', or 'bash <command>'. Available tools: {names}."
            ))],
        }
    }
}

impl MockProvider {
    fn tool_call(&mut self, text: &str, name: &str, args: Vec<(String, String)>) -> Message {
        self.counter += 1;
        Message {
            role: "assistant".to_string(),
            content: vec![
                ContentItem::Text(text.to_string()),
                ContentItem::ToolCall {
                    id: format!("call_{}", self.counter),
                    name: name.to_string(),
                    arguments: args,
                },
            ],
        }
    }
}

pub fn extract_text(message: &Message) -> String {
    message
        .content
        .iter()
        .filter_map(|c| match c {
            ContentItem::Text(t) => Some(t.clone()),
            _ => None,
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn has_tool(tools: &[ToolSpec], name: &str) -> bool {
    tools.iter().any(|t| t.name == name)
}
